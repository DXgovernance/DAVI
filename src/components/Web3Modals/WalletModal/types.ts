export interface WalletModalProps {
  isOpen: boolean;
  onClose: () => void;
}
