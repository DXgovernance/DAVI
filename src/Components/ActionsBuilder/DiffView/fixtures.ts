export const mockCodeOriginal = `
  {
    "colors": {
      "text": "red",
      "primary": "#295FF4",
      "secondary": "#181B1B",
      "red": "#E75C5C",
      "orange": "#D17A06",
      "proposalText": {
        "lightGrey": "#BDC0C7",
        "grey": "orange"
      },
    }
  }
`;

export const mockCodeUpdated = `
  {
    "colors": {
      "text": "#fff",
      "primary": "#295FF4",
      "secondary": "#181B1B",
      "red": "#E75C5C",
      "orange": "#D17A06",
      "proposalText": {
        "grey": "#A1A6B0",
        "lightGrey": "#BDC0C7"
      },
    }
  }
`;
