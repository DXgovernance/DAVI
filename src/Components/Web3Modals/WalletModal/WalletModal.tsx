import { useState } from 'react';
import { useTranslation } from 'react-i18next';
import { Modal } from 'Components/Primitives/Modal';
import {
  BackIcon,
  ButtonContainer,
  ConnectionError,
  Container,
  TransactionsList,
  TransactionsListHeading,
} from './WalletModal.styled';
import { WalletModalProps } from './types';
import {
  useAccount,
  useConnect,
  useDisconnect,
  useNetwork,
  useSwitchNetwork,
} from 'wagmi';
import { Option, Transaction, WalletInfoBox } from '../components';
import { getIcon, isReadOnly } from 'provider/wallets';
import { useTransactions } from 'contexts/Guilds';
import { Divider } from 'old-components/Guilds/common/Divider';
import { Button } from 'old-components/Guilds/common/Button';
import { READ_ONLY_CONNECTOR_ID } from 'provider/ReadOnlyConnector';

const WalletModal: React.FC<WalletModalProps> = ({ isOpen, onClose }) => {
  const { t } = useTranslation();
  const [isWalletListActive, setIsWalletsListActive] = useState(false);
  const { chain } = useNetwork();
  const { chains, switchNetwork } = useSwitchNetwork();
  const { transactions, clearAllTransactions } = useTransactions();
  const { disconnect } = useDisconnect();
  const {
    connector: activeConnector,
    isConnecting,
    isConnected,
  } = useAccount();
  const { connectors, connect, error: connectionError } = useConnect();

  function getOptions() {
    return connectors
      .filter(connector => connector.id !== READ_ONLY_CONNECTOR_ID)
      .map(connector => {
        return (
          <Option
            dataTestId={`wallet-option-${connector.id}`}
            disabled={!connector.ready}
            key={connector.id}
            onClick={() =>
              activeConnector !== connector && connect({ connector })
            }
            icon={getIcon(connector.id, connector.name)}
            header={connector.name}
            active={activeConnector === connector && isConnected}
            loading={activeConnector === connector && isConnecting}
          />
        );
      });
  }

  function getModalContent() {
    if (isConnected && chain.unsupported) {
      return <Container>{t('pleaseSwitchNetwork')}</Container>;
    }

    if (!isConnected || isWalletListActive || isReadOnly(activeConnector)) {
      return (
        <Container>
          {getOptions()}

          {connectionError && (
            <ConnectionError>{connectionError.message}</ConnectionError>
          )}
        </Container>
      );
    }

    const recentTransactions =
      transactions &&
      transactions
        .sort((tx1, tx2) => tx2.addedTime - tx1.addedTime)
        .slice(0, 5);
    return (
      <>
        <WalletInfoBox openOptions={() => setIsWalletsListActive(true)} />
        <Divider />
        <TransactionsList>
          {recentTransactions?.length === 0 ? (
            <TransactionsListHeading>
              {t('yourTransactionsWillAppearHere')}
            </TransactionsListHeading>
          ) : (
            <>
              <TransactionsListHeading>
                {t('recentTransactions')}
              </TransactionsListHeading>
              <Divider />
              {recentTransactions?.map(transaction => (
                <Transaction transaction={transaction} key={transaction.hash} />
              ))}
              {recentTransactions?.length > 0 && (
                <ButtonContainer>
                  <Button onClick={clearAllTransactions}>
                    {t('clearAll')}
                  </Button>
                </ButtonContainer>
              )}
            </>
          )}
        </TransactionsList>
      </>
    );
  }

  const getHeader = () => {
    if (isConnected && chain.unsupported) {
      return t('unsupportedNetwork');
    }

    if (isConnected && isWalletListActive && !isReadOnly(activeConnector)) {
      return (
        <BackIcon
          onClick={() => {
            setIsWalletsListActive(false);
          }}
        />
      );
    }

    return isConnected && !isReadOnly(activeConnector)
      ? t('account')
      : t('connectToAWallet');
  };

  const getPrimaryAction = () => {
    if (isConnected && isWalletListActive && !isReadOnly(activeConnector))
      return () => disconnect();

    if (isConnected && chain.unsupported && switchNetwork) {
      const firstSupported = chains && chains?.length > 0 ? chains[0] : null;
      return () => switchNetwork(firstSupported.id);
    }

    return null;
  };

  const getPrimaryActionLabel = () => {
    if (isConnected && isWalletListActive && !isReadOnly(activeConnector))
      return t('disconnect');

    if (isConnected && chain.unsupported) {
      const firstSupported = chains && chains?.length > 0 ? chains[0] : null;

      return t('switchNetworkTo', { chainName: firstSupported?.name });
    }

    return null;
  };

  return (
    <Modal
      dataTestId="wallet-modal"
      header={getHeader()}
      isOpen={isOpen}
      onDismiss={onClose}
      onConfirm={getPrimaryAction()}
      confirmText={getPrimaryActionLabel()}
      maxWidth={450}
    >
      {getModalContent()}
    </Modal>
  );
};

export default WalletModal;
