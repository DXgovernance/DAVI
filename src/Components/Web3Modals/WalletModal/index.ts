export { default as WalletModal } from './WalletModal';
