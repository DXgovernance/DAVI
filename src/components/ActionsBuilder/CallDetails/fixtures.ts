import {
  ApproveSendTokens,
  DecodedCall,
  SupportedAction,
} from 'components/ActionsBuilder/types';
import ERC20 from 'contracts/ERC20.json';
import { BigNumber, utils } from 'ethers';

const ERC20Contract = new utils.Interface(ERC20.abi);

export const emptyDecodedCallMock: DecodedCall = {
  from: '0x0000000000000000000000000000000000000000',
  callType: SupportedAction.ERC20_TRANSFER,
  function: ERC20Contract.getFunction('transfer'),
  to: '',
  value: BigNumber.from(0),
  args: {
    _to: '',
    _value: BigNumber.from(0),
  },
};

export const decodedCallMock: DecodedCall = {
  from: '0x0000000000000000000000000000000000000000',
  callType: SupportedAction.ERC20_TRANSFER,
  function: ERC20Contract.getFunction('transfer'),
  to: '0x95a223299319022a842d0dfe4851c145a2f615b9',
  value: BigNumber.from(0),
  args: {
    _to: '0x0000000000000000000000000000000000000000',
    _value: BigNumber.from(1000),
  },
};

export const approvetokenSpendingMock: ApproveSendTokens = {
  callType: SupportedAction.GENERIC_CALL,
  from: '0x0000000000000000000000000000000000000000',
  to: '0x0000000000000000000000000000000000000000',
  value: BigNumber.from(0),
  function: null,
  args: {},
  amount: BigNumber.from(1000),
  token: '0x3f943f38b2fbe1ee5daf0516cecfe4e0f8734351',
};
