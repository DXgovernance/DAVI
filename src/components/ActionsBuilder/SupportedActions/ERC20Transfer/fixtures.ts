import { DecodedCall, SupportedAction } from 'components/ActionsBuilder/types';
import ERC20ABI from 'contracts/ERC20.json';
import { BigNumber, utils } from 'ethers';

const ERC20Contract = new utils.Interface(ERC20ABI.abi);

export const erc20TransferEmptyDecodedCallMock: DecodedCall = {
  from: '0x0000000000000000000000000000000000000000',
  callType: SupportedAction.ERC20_TRANSFER,
  function: ERC20Contract.getFunction('transfer'),
  to: '',
  value: BigNumber.from(0),
  args: {
    _to: '',
    _value: BigNumber.from(0),
  },
};

export const erc20TransferDecodedCallMock: DecodedCall = {
  from: '0x0000000000000000000000000000000000000000',
  callType: SupportedAction.ERC20_TRANSFER,
  function: ERC20Contract.getFunction('transfer'),
  to: '0xfFb1cd0F95368DDd06D556161c5D3d9f0f4Fe6d2',
  value: BigNumber.from(0),
  args: {
    _to: '0xaF8eB8C3A5d9d900AA0B98e3Df0bcC17d3C5F698',
    _value: BigNumber.from('5000000000000000000'),
  },
};
