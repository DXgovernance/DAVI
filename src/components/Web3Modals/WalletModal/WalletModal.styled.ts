import styled from 'styled-components';
import { FiArrowLeft } from 'react-icons/fi';
import { Heading } from 'components/Primitives/Typography';

export const Container = styled.div`
  margin: 2rem;
`;

export const ConnectionError = styled.div`
  text-align: center;
  color: ${({ theme }) => theme.colors.red};
`;

export const BackIcon = styled(FiArrowLeft)`
  height: 1.5rem;
  width: 1.5rem;
  cursor: pointer;
  margin: 0;
  padding: 0;

  &:hover {
    color: ${({ theme }) => theme.colors.muted};
  }
`;

export const ButtonContainer = styled.div`
  margin: 1.5rem;
  text-align: center;
`;

export const TransactionsList = styled.div`
  margin: 1.5rem 0;
`;

export const TransactionsListHeading = styled(Heading)`
  margin-left: 1.5rem;
  margin-right: 1.5rem;
  font-size: ${({ theme }) => theme.fontSizes.body};
  color: ${({ theme }) => theme.colors.proposalText.grey};
`;
