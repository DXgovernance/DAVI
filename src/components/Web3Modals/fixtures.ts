import { ZERO_ADDRESS } from 'utils';
import { Chain } from 'wagmi';

export const mockChain: Chain = {
  id: 123456,
  name: 'Random Chain',
  network: 'random',
  rpcUrls: {
    default: 'http://localhost:8545',
  },
};

export const mockAccount = {
  address: ZERO_ADDRESS,
  isConnected: false,
  isConnecting: false,
  isDisconnected: false,
  isReconnecting: false,
  status: 'disconnected',
  connector: null,
};
