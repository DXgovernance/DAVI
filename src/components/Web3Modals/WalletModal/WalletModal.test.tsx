import { WalletModal } from './WalletModal';
import { render } from 'utils/tests';
import { mockChain } from '../fixtures';

jest.mock('contexts/Guilds', () => ({
  useTransactions: () => [],
}));

jest.mock('provider/ReadOnlyConnector', () => ({
  READ_ONLY_CONNECTOR_ID: 'readOnly',
}));

jest.mock('wagmi', () => ({
  chain: {},
  useAccount: () => ({ isConnected: true }),
  useNetwork: () => ({ chain: mockChain, chains: [mockChain] }),
  useSwitchNetwork: () => ({ switchNetwork: jest.fn() }),
  useConnect: () => ({ connect: jest.fn(), connectors: [] }),
  useDisconnect: () => ({ disconnect: jest.fn() }),
}));

describe('WalletModal', () => {
  it('Should match snapshot', () => {
    console.error = jest.fn();
    const { container } = render(
      <WalletModal isOpen={true} onClose={() => jest.fn()} />,
      {
        container: document.body,
      }
    );
    expect(container).toMatchSnapshot();
  });
});
