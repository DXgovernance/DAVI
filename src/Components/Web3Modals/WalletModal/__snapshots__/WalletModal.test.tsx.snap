// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WalletModal Should match snapshot 1`] = `
<body>
  <div
    data-testid="wallet-modal"
  >
    <div
      class="sc-aemIz iegKhJ backdropModal"
    >
      <div
        class="sc-jHVedQ jaSA-db"
      >
        <div
          class="sc-jONnzC bZnDLt"
        >
          
           
          <div
            class="sc-hlTwrV hTyPNj"
          >
            <h2
              class="sc-dQoVA sc-bTvSiT fvNDvx iJsJQz"
            >
              account
            </h2>
            <svg
              class="sc-kIeSZW ktHRrf"
              fill="none"
              height="1em"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <line
                x1="18"
                x2="6"
                y1="6"
                y2="18"
              />
              <line
                x1="6"
                x2="18"
                y1="6"
                y2="18"
              />
            </svg>
          </div>
          <div
            class="sc-dtwnrN gnLHYk"
          >
            <div
              class="sc-GTVdH QKZBv"
            >
              <a
                href="https://metamask.io/"
              >
                <button
                  class="sc-bdfBQB sc-jNMcJZ cAORUT jctxDF"
                  data-testid=""
                >
                  <div
                    class="sc-dOSRxR bvcoLH"
                  >
                    installMetamask
                  </div>
                  <div
                    class="sc-bBrNTY dkBKFH"
                  >
                    <img
                      alt="Icon"
                      src="metamask.png"
                    />
                  </div>
                </button>
              </a>
              <button
                class="sc-bdfBQB sc-jNMcJZ cAORUT jctxDF"
                data-testid="wallet-option-WalletConnect"
              >
                <div
                  class="sc-dOSRxR bvcoLH"
                >
                  WalletConnect
                </div>
                <div
                  class="sc-bBrNTY dkBKFH"
                >
                  <img
                    alt="Icon"
                    src="walletconnect.png"
                  />
                </div>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
   
   
</body>
`;
